container/nginx/njs/auth.js

```diff
@@ -0,0 +1,73 @@
+import crypto from "crypto";
+
+const ipfsRegex = /^\/ipfs\/(\w+)(\/?.*)/;
+
+function isAllowedRequest(req) {
+  const matches = req.uri.match(ipfsRegex);
+  if (!matches) {
+    return req.internalRedirect("@node_backend");
+  }
+  const cid = matches[1];
+
+  if (isBadBitsCid(cid)) {
+    return req.return(410);
+  }
+
+  if (!isAllowedDomain(req)) {
+    return req.return(403);
+  }
+
+  req.internalRedirect("@node_backend");
+}
+
+// TODO implement matching CID paths
+// TODO convert CID v0 to CID v1
+// implementation ref: https://github.com/protocol/bifrost-infra/blob/af46340bd830728b38a0ea632ca517d04277f78c/ansible/roles/nginx_conf_denylist/files/lua/helpers.lua#L80
+function isBadBitsCid(cid) {
+  // check if root hash(`CID/`) is blocked via denylist.json
+  const hashedCID = crypto
+    .createHash("sha256")
+    .update(cid + "/")
+    .digest("hex");
+
+  /* eslint-disable-next-line no-undef */
+  return hashedCID in denylist;
+}
+
+function isAllowedDomain(req) {
+  const allowListStr = req.variables.jwt_claim_allow_list;
+  // Only browser requests are allowed for now.
+  const requestOrigin = req.variables.http_origin;
+  if (!allowListStr || !requestOrigin) {
```

Comment:
Requests without an `Origin` header will fail. Is this what we want?

Browsers will automatically send an `Origin` header in most cases. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin. For Saturn, all requests in the browser should be cross origin fetches, so the `Origin` header should be added.

Reply:
kind of nuetral on this -- what does it get us? I would generally expect [*] to allow empty origin, but not particularly strong on this.

Reply:
one worry -- what about non-browser requests? we're already working on native clients for games. they can of course still set origins.

Reply:
> kind of nuetral on this -- what does it get us? 

I checked with @tchardin and this seems to be the current requirement, but I'll admit it feels weird. 

> I would generally expect [*] to allow empty origin, but not particularly strong on this.

I had the same thought, I can add this.

> one worry -- what about non-browser requests? we're already working on native clients for games. they can of course still set origins.

Origin header checking is involved in both JWT creation and verification, it's very browser centric atm. The non-browser auth flow is still TBD I guess.

Reply:
* the current flow for lab week is for browser clients only. we're purposefully not onboarding any customers beyond browsers for the private beta launch

* we 100% need to support non-browser clients in the future, and making every client, browser or not, set an `Origin` header is one way to do it, but my gut strongly says not the best way. its hacky and non-obvious to overload `Origin`

* in my mind, the right/better approach is to update the JWT in the future when it's handed out to non-browser clients -- those that don't send an `Origin` header. then the l1 could:
  * first check the JWT to see what kind of JWT it is. is it one for a website with an allowlist of domains? if so, check for the `Origin` header and test for a match 👍 
  * if the JWT doesn't have an allowlist of domains, its for an app. like a native game. and the auth check will be different 
  
* we can change which claims are included in the JWT's claims payloads and they can differ depending on what kind of client is making the request

* for native apps, perhaps instead of an `Origin` header, we could use a different, custom header, like `Saturn-App-Id`, that gets set by the smart client in the native app, and this `Saturn-App-Id` is corroborated with the custom JWT claim we add for native apps. but, as of now, this hasn't been designed

either way, i dont think we need to solve this problem in the future once we get a better idea of non-browser use cases. fundamentally, the entire point of having auth is so we can track, meter, and bill customers for their usage. and, where possible, prevent abuse by say game B using game A's saturn account to distribute content. the latter becomes harder when every client can 'lie' easily, which isn't possible at scale across a whole website with the `Origin` header. this requires some more thinking and exploring how saturn will be used by non-browser clients